lib/transmuxer.js

```diff
@@ -54,52 +54,101 @@ mp4 = muxjs.mp4;
  * Splits an incoming stream of binary data into MPEG-2 Transport
  * Stream packets.
  */
+
 TransportPacketStream = function() {
   var
     buffer = new Uint8Array(MP2T_PACKET_LENGTH),
-    end = 0;
+    bytesInBuffer = 0;
 
   TransportPacketStream.prototype.init.call(this);
 
-  /**
-   * Deliver new bytes to the stream.
-   */
-  this.push = function(bytes) {
-    var remaining, i;
-
-    // clear out any partial packets in the buffer
-    if (end > 0) {
-      remaining = MP2T_PACKET_LENGTH - end;
-      buffer.set(bytes.subarray(0, remaining), end);
+   // Deliver new bytes to the stream.
 
-      // we still didn't write out a complete packet
-      if (bytes.byteLength < remaining) {
-        end += bytes.byteLength;
-        return;
-      }
+  this.push = function(bytes) {
+    var
+      bytesRemaining,
+      i = 0;
 
-      bytes = bytes.subarray(remaining);
-      end = 0;
-      this.trigger('data', buffer);
-    }
+    bytes = this.makePacketFromBufferedData_(bytes);
 
-    // if less than a single packet is available, buffer it up for later
-    if (bytes.byteLength < MP2T_PACKET_LENGTH) {
-      buffer.set(bytes.subarray(i), end);
-      end += bytes.byteLength;
+    // If bytes is undefined, then we don't have enough data for one
+    // complete packet so just return and wait for more
+    if (!bytes) {
       return;
     }
-    // parse out all the completed packets
-    i = 0;
-    do {
+
+    bytesRemaining = bytes.byteLength;
+
+    // Parse out all the completed packets
+    while (i + MP2T_PACKET_LENGTH <= bytes.byteLength) {
       this.trigger('data', bytes.subarray(i, i + MP2T_PACKET_LENGTH));
       i += MP2T_PACKET_LENGTH;
-      remaining = bytes.byteLength - i;
-    } while (i < bytes.byteLength && remaining >= MP2T_PACKET_LENGTH);
-    // buffer any partial packets left over
-    if (remaining > 0) {
+      bytesRemaining -= MP2T_PACKET_LENGTH;
+    }
+
+    // Buffer any partial packets left over
+    if (bytesRemaining > 0) {
       buffer.set(bytes.subarray(i));
-      end = remaining;
+      bytesInBuffer = bytesRemaining;
+    }
+  };
+
+  // Attempt to complete any partial packets from data that is in the
+  // `buffer` and new data we've just recieved
+  this.makePacketFromBufferedData_ = function (bytes) {
+    var bytesNeeded, bytesLeft;
+
+    while (true) {
```

Comment:
Discussed offline: @imbcmdth and I agreed we need to try to take a slightly different tack than this implementation.